File: C/221.c
//暴力算法，对每个值为1的元素，找它为左上角的最大正方形

#define Min(a, b) ((a) < (b) ? (a) : (b))
int FindMax(char **matrix, int i, int j, int matrixSize, int *matrixColSize)
{
    int ans = 1;
    int currentMaxSize = Min(matrixSize - i, matrixColSize[i] - j);
    for (int k = 1; k < currentMaxSize; k++) {
        int flag = 1;
        if (matrix[i + k][j + k] == '0')
            break;
        for (int m = 0; m < k; m++) {
            if (matrix[i + k][j + m] == '0' || matrix[i + m][j + k] == '0') {
                flag = 0;
                break;
            }
        }
        if (flag == 1)
            ans = (k + 1) > ans ? (k + 1) : ans;
        else 
            break;
    }
    return ans;
}

int maximalSquare(char** matrix, int matrixSize, int* matrixColSize){
    int length = 0, tmp;
    for (int i = 0; i < matrixSize; i++) {
        for (int j = 0; j < matrixColSize[i]; j++) {
            if (matrix[i][j] == '1') {
                tmp = FindMax(matrix, i, j, matrixSize, matrixColSize);
                if (tmp > length)
                    length = tmp;
            }
        }
    }   
    return length * length;
}

//动态规划，关键是dp[i][j] = Min(Min(dp[i - 1][j], dp[i][j - 1]), dp[i - 1][j - 1]) + 1;  dp[i][j]是(i，j)作为右下角时的最大正方形
#define Min(a, b) ((a) < (b) ? (a) : (b))

int maximalSquare(char** matrix, int matrixSize, int* matrixColSize){
    int length = 0;
    int **dp;
    dp = (int **)malloc(sizeof(int *) * matrixSize);
    for (int i = 0; i < matrixSize; i++)
        dp[i] = (int *)malloc(sizeof(int) * matrixColSize[0]);
    for (int i = 0; i < matrixSize; i++) {
        for (int j = 0; j < matrixColSize[0]; j++)
            dp[i][j] = 0;
    }
    for (int i = 0; i < matrixSize; i++) {
        for (int j = 0; j < matrixColSize[i]; j++) {
            if (matrix[i][j] == '1') {
                if (i == 0 || j == 0)
                    dp[i][j] = 1;
                else 
                    dp[i][j] = Min(Min(dp[i - 1][j], dp[i][j - 1]), dp[i - 1][j - 1]) + 1;
                if (dp[i][j] > length)
                    length = dp[i][j];
            }
        }
    }
    return length * length;
}

File: Java/221.java
//用暴力法的时间居然比动态规划短，我吐了

class Solution {
    public int maximalSquare(char[][] matrix) {
        int res = 0;
        int m = matrix.length;
        int n = matrix[0].length;
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                int limit = Math.min(m - i, n - j);
                if (matrix[i][j] == '1' && limit * limit > res) {
                    res = Math.max(res, 1);
                    int length;
                    boolean flag = true;
                    for (length = 1; length < limit; length++) {
                        for (int x = 0; x < length; x++) {
                            if (matrix[i + length][j + x] == '0' || matrix[i + x][j + length] == '0') {
                                flag = false;
                                break;
                            }
                        }
                        flag = flag && matrix[i + length][j + length] == '1';
                        if (!flag) {
                            break;
                        } else {
                            res = Math.max(res, (length + 1) * (length + 1));
                        }
                    }
                }
            }
        }
        return res;
    }
}

class Solution {
    public int maximalSquare(char[][] matrix) {
        int res = 0;
        int m = matrix.length;
        int n = matrix[0].length;
        int[][] dp = new int[m][n];
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (matrix[i][j] == '1') {
                    if (i > 0 && j > 0) {
                        dp[i][j] = Math.min(Math.min(dp[i - 1][j], dp[i - 1][j - 1]), dp[i][j - 1]) + 1;
                    } else {
                        dp[i][j] = 1;
                    }
                    res = Math.max(res, dp[i][j]);
                } else {
                    dp[i][j] = 0;
                }
            }
        }
        return res * res;
    }
    
}